src/storage/CommonUtils.h

```diff
@@ -59,6 +60,10 @@ using VertexCache = ConcurrentLRUCache<std::pair<VertexID, TagID>, std::string>;
 using IndexKey    = std::tuple<GraphSpaceID, PartitionID>;
 using IndexGuard  = folly::ConcurrentHashMap<IndexKey, IndexState>;
 
+using VMLI = std::tuple<GraphSpaceID, PartitionID, TagID, VertexID>;
```

Comment:
Have you try for string key? I don't know how much difference with tuple and string in ConcurrentHashMap benchmark.

Reply:
> Have you try for string key? I don't know how much difference with tuple and string in ConcurrentHashMap benchmark.

Your means, instead it to tuple<spaceId, string_key>?
This has been tried before in a draft version and has not improved performance. The current implementation has two advantages, 1) easier access to conflict item (Vid, src, dst, edgeType, rank) for error report. 2) more intuitive, edge version(1 byte) are easier to control.

Reply:
Ok.